File: python/coin_change.py
class Solution:
    # # brute force
    # def coinChange(self, coins: List[int], amount: int) -> int:
    #     # cannot use coin path to make original amount
    #     if amount < 0:
    #         return -1
    #     # can use coin path to make original amount,
    #     # since it takes 0 coins to make 0 amount
    #     if amount == 0:
    #         return 0

    #     # min coin count needed to make amount up to now
    #     minCoinCountSoFar = float('inf')

    #     # find min coin combination needed to make current amount
    #     for coin in coins:
    #         coinCount = self.coinChange(coins, amount - coin)

    #         if coinCount == -1:
    #             continue

    #         minCoinCountSoFar = min(coinCount, minCoinCountSoFar)

    #     # if no valid coin paths to reach current amount
    #     if minCoinCountSoFar == float('inf'):
    #         return -1

    #     # add 1 coin to account for lost coin during subtraction
    #     return minCoinCountSoFar + 1

    # # optimized (memoized)
    # def coinChange(self, coins: List[int], amount: int) -> int:
    #     return self.countCoinChange(coins, amount, {})

    # def countCoinChange(self, coins: List[int], amount: int, minChangeByAmount: dict[int, int]) -> int:
    #     # cannot use coin path to make original amount
    #     if amount < 0:
    #         return -1
    #     # can use coin path to make original amount,
    #     # since it takes 0 coins to make 0 amount
    #     if amount == 0:
    #         return 0

    #     # try hitting cache for min coin change at amount
    #     if amount in minChangeByAmount:
    #         return minChangeByAmount[amount]

    #     # min coin count needed to make amount up to now
    #     minCoinCountSoFar = float('inf')

    #     # find min coin combination needed to make current amount
    #     for coin in coins:
    #         coinCount = self.countCoinChange(coins, amount - coin, minChangeByAmount)

    #         if coinCount == -1:
    #             continue

    #         minCoinCountSoFar = min(coinCount, minCoinCountSoFar)

    #     # if no valid coin paths to reach current amount
    #     if minCoinCountSoFar == float('inf'):
    #         minChangeByAmount[amount] = -1
    #         return -1

    #     # update cache with min coin change at amount
    #     minChangeByAmount[amount] = minCoinCountSoFar + 1

    #     # add 1 coin to account for lost coin during subtraction
    #     return minCoinCountSoFar + 1

    # optimal bottom up dp
    def coinChange(self, coins: List[int], amount: int) -> int:
        minSteps = []
        minSteps.append(0) # it takes 0 steps to get to 0
        
        for currAmount in range(1, amount + 1):
            # placeholder
            minSteps.append(float("inf"))
            for coin in coins:
                # cannot subtract to get subproblem
                if coin > currAmount:
                    continue
                # find subproblem (the difference by subtracting the coin)
                diff = currAmount - coin
                # least num of steps to get currAmount is the smallest
                # between least num of steps to get subproblem + 1 (to account                     # for coin that gets us to subproblem) or current num of steps                     # already
                minSteps[currAmount] = min(minSteps[diff] + 1, minSteps[currAmount])

        # if amount cannot be made up of the coins 
        # (inf is propagated from subproblems, or by all and amount)
        if minSteps[amount] == float("inf"):
            return -1

        # min num of steps/coins to get amount
        return minSteps[amount]

File: cpp/coin_change.cpp
class Solution {
public:
    // bottom up dp approach
    int coinChange(vector<int> &coins, int amount) {
        // build up solution of minimum coins needed to make each amount, default to not
        // possible initially and zero coins trivially needed to make amount of 0
        vector<int> minCoinsByAmount(amount + 1, INT_MAX);
        minCoinsByAmount[0] = 0;

        for (int curAmount = 1; curAmount <= amount; ++curAmount) {
            for (int &coin : coins) {
                // cannot make current amount with current coin (exceeds amount)
                if (curAmount - coin < 0) {
                    continue;
                }

                // minimum number of coins (steps) to make current amount depends on the
                // minimum number of coins to make amount one coin denomination away if
                // possible (+ 1 coin including current)
                minCoinsByAmount[curAmount] = min(minCoinsByAmount[curAmount - coin], minCoinsByAmount[curAmount]);
            }

            // only add back coin to make current amount possible to make previous amount
            if (minCoinsByAmount[curAmount] < INT_MAX) {
                ++minCoinsByAmount[curAmount];
            }
        }

        int minCoins = minCoinsByAmount[amount];
        return minCoins == INT_MAX ? -1 : minCoins;
    }

    // // top down dp approach
    // int coinChange(vector<int> &coins, int amount) {
    //     // cache minimum number of coins needed to make each unique amount
    //     unordered_map<int, int> cache;
    //     int minCoins = minRequiredCoins(coins, amount, cache);
    //     return minCoins == INT_MAX ? -1 : minCoins;
    // }

    // int minRequiredCoins(vector<int> &coins, int amount, unordered_map<int, int> &cache) {
    //     // previous coin used does not make exact change (exceeds amount)
    //     if (amount < 0) {
    //         return INT_MAX;
    //     }
    //     // zero coins trivially needed to make amount of 0
    //     if (amount == 0) {
    //         return 0;
    //     }

    //     // previously computed minimum number of coins to make current amount
    //     if (cache.find(amount) != cache.end()) {
    //         return cache[amount];
    //     }

    //     // minimum number of coins to make current amount depends on minimum coins to
    //     // make amount one coin denomination away (+ 1 coin including current)
    //     int minCoinsNextAmount = INT_MAX;
    //     for (int &coin : coins) {
    //         minCoinsNextAmount = min(minCoinsNextAmount, minRequiredCoins(coins, amount - coin, cache));
    //     }

    //     // only add back coin to make current amount possible to make previous amount
    //     if (minCoinsNextAmount < INT_MAX) {
    //         ++minCoinsNextAmount;
    //     }
    //     cache[amount] = minCoinsNextAmount;

    //     return minCoinsNextAmount;
    // }
};
